sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala

```diff
@@ -120,7 +120,25 @@ case class Generate(
 
 case class Filter(condition: Expression, child: LogicalPlan)
   extends UnaryNode with PredicateHelper {
-  override def output: Seq[Attribute] = child.output
+
+  // The columns that will filtered out by `IsNotNull` could be considered as not nullable.
+  private lazy val notNullAttributes = splitConjunctivePredicates(condition).flatMap {
+    case isnotnull @ IsNotNull(a)
+        if isNullIntolerant(a) && a.references.subsetOf(child.outputSet) =>
+      isnotnull.references.map(_.exprId)
+    case _ =>
+      Seq.empty[ExprId]
+  }.toSet
+
+  override def output: Seq[Attribute] = {
+    child.output.map { a =>
+      if (a.nullable && notNullAttributes.contains(a.exprId)) {
```

Comment:
The same issue also applies to the other nodes. Filter is just one of them, if we use constraints to determine whether they are nullable or not

Reply:
oh, yea. Certainly, each `output` in logical plans always should determine nullability  based on their own `constraints`.